File: packs/evidence-standing-pack/qualification/drivers/driver.ts
// Golden-vector driver (bun). Usage: bun driver.ts <golden.vec>; imports generated src/es/chain.ts.
import { readFileSync } from "node:fs";
import { appendPending, appendOutcome, seal, unpaired, verify, digest, canonical, GENESIS, type Chain, type Entry } from "./src/es/chain.ts";

let chain: Chain = [];
for (const raw of readFileSync(process.argv[2], "utf8").split("\n")) {
  if (!raw || raw.startsWith("#")) continue;
  const p = raw.split("|");
  if (raw.startsWith("case ")) { chain = []; console.log("case", raw.slice(5)); }
  else if (p[0] === "pending" || p[0] === "outcome" || p[0] === "seal") {
    try {
      const e = p[0] === "pending" ? appendPending(chain, p[1], p[2], p[3]) : p[0] === "outcome" ? appendOutcome(chain, p[1], p[2], p[3], p[4]) : seal(chain, p[1], p[2], p[3]);
      console.log("ok", e.hash);
    } catch { console.log("err"); }
  } else if (p[0] === "forge") {
    const i = Number(p[6]);
    const e: Entry = { entry_id: p[1], parent_hash: chain.length ? chain[chain.length - 1].hash : GENESIS, phase: p[2], standing: p[3], subject: p[4], action: p[5], pending_ref: i >= 0 ? chain[i].hash : "", hash: "" };
    e.hash = digest(canonical(e as unknown as Record<string, unknown>));
    chain.push(e);
    console.log("ok", e.hash);
  } else if (p[0] === "unpaired") console.log("unpaired", unpaired(chain).join(","));
  else if (p[0] === "verify") console.log("verify", verify(chain) ? "true" : "false");
  else if (p[0] === "tamper") { chain[0].subject = "tampered"; console.log("tampered"); }
}

File: packs/evidence-standing-pack/qualification/drivers/driver.exs
# Golden-vector driver (elixir). Usage: elixir -r lib/es/chain.ex driver.exs <golden.vec>
defmodule Driver do
  def run(path) do
    path |> File.read!() |> String.split("\n") |> Enum.reduce([], &step/2)
  end

  defp step("", chain), do: chain
  defp step("#" <> _, chain), do: chain
  defp step("case " <> name, _chain) do
    IO.puts("case " <> name)
    []
  end
  defp step(line, chain) do
    case String.split(line, "|") do
      ["pending", id, su, ac] -> res(Es.Chain.append_pending(chain, id, su, ac), chain)
      ["outcome", id, st, su, ac] -> res(Es.Chain.append_outcome(chain, id, st, su, ac), chain)
      ["seal", id, st, su] -> res(Es.Chain.seal(chain, id, st, su), chain)
      ["forge", id, ph, st, su, ac, i] ->
        i = String.to_integer(i)
        ref = if i >= 0, do: Enum.at(chain, i).hash, else: ""
        parent = case List.last(chain) do nil -> Es.Chain.genesis(); p -> p.hash end
        e = %{entry_id: id, parent_hash: parent, phase: ph, standing: st, subject: su, action: ac, pending_ref: ref, seal: false}
        e = Map.put(e, :hash, Es.Chain.digest(Es.Chain.canonical(e)))
        IO.puts("ok " <> e.hash)
        chain ++ [e]
      ["unpaired"] ->
        IO.puts("unpaired " <> Enum.join(Es.Chain.unpaired(chain), ","))
        chain
      ["verify"] ->
        IO.puts("verify " <> to_string(Es.Chain.verify(chain)))
        chain
      ["tamper"] ->
        IO.puts("tampered")
        [%{hd(chain) | subject: "tampered"} | tl(chain)]
    end
  end

  defp res({:ok, c}, _old) do
    IO.puts("ok " <> List.last(c).hash)
    c
  end
  defp res({:error, _}, old) do
    IO.puts("err")
    old
  end
end

Driver.run(hd(System.argv()))
